blocklib/timing/include/TimingSource.hpp

```diff
@@ -0,0 +1,484 @@
+    A<bool, "verbose console", Doc<"For debugging">>                                                                                                                                                                            verbose_console = false;
+    // TODO: enable/disable publishing on input port changes -> For now everything is published, add in follow-up PR
+
+    GR_MAKE_REFLECTABLE(TimingSource, out, ctxInfo, event_actions, sample_rate, timing_device, max_delay, verbose_console);
+
+    using ConditionsType = std::map<std::string, std::map<std::string, std::variant<std::shared_ptr<saftlib::OutputCondition_Proxy>, std::shared_ptr<saftlib::SoftwareCondition_Proxy>>>>;
+    Timing _timing;
+    using EventReaderType         = decltype(_timing.snooped.new_reader());
+    EventReaderType _event_reader = _timing.snooped.new_reader();
+    ConditionsType  _conditionProxies;
+    TimePoint       _startTime;
+    std::size_t     _publishedSamples = 0;
+    std::uint8_t    _lastOutputState  = 0;
+    std::uint8_t    _outputState      = 0;
+    std::uint8_t    _nextOutputState  = 0;
+
+    void start() {
+        if (verbose_console) {
+            fmt::println("starting {}", this->name);
+        }
+        _timing.saftAppName = fmt::format("{}_{}", this->unique_name | std::views::filter([](auto c) { return std::isalnum(c); }) | std::ranges::to<std::string>(), getpid());
+        _timing.snoopID     = 0xffffffffffffffffull; // Workaround: make the default condition of _timing not listen to anything
+        _timing.snoopMask   = 0xffffffffffffffffull;
+        _timing.initialize();
+        updateEventTriggers();
+        _startTime = TimePoint(std::chrono::nanoseconds(_timing.currentTimeTAI()));
+    }
+
+    void stop() {
+        if (verbose_console) {
+            fmt::println("stop {}", this->name);
+        }
+        this->requestStop();
+    }
+
+    static std::pair<std::uint64_t, std::uint64_t> parseFilter(const std::string& newFilter) {
+        std::vector<std::string_view> filterTokens = std::views::split(newFilter, ":"sv) | std::views::transform([](auto r) { return std::string_view(r.begin(), r.end()); }) | std::ranges::to<std::vector>();
+        // initialize with the fixed format identifier
+        uint64_t filter = 0x1000000000000000UL;
+        uint64_t mask   = 0xf000000000000000UL;
+        // timing group ID, given as string or numeric value
+        if (filterTokens.empty()) {
+            return {filter, mask};
+        }
+        if (!filterTokens[0].empty()) {
+            char*    end            = const_cast<char*>(filterTokens[0].end());
+            uint64_t numericGroupId = std::strtoull(filterTokens[0].begin(), &end, 0);
+            if (end == filterTokens[0].end()) {
```

Comment:
I could be better to use `std::from_chars`. A `std::string_view is not guaranteed to be null-terminated but `std::strtoull`requires it. 

```cpp
uint64_t numericGroupId;
auto [ptr, err] = std::from_chars(str.data(), str.data() + str.size(), numericGroupId);
if (err == std::errc{} && ptr == str.data() + str.size()) {
    // conversion is successful
}
```

Reply:
good catch, I had assumed that the existing value of end would be respected, but it seems to solely use it to return how far it parsed :+1:,  will fix.